### PR TITLE
Terraform 0.12 support #11

### PR DESCRIPTION
**Pull request by ingwarr**
Fixes https://github.com/terraform-google-modules/terraform-google-vpn/issues/11

### CODE REVIEW

**Comment by ingwarr** [comment]
Re-tested with docker

File: README.md [hunk 1]
```diff
@@ -104,7 +104,7 @@ Please refer the /outputs.tf file for the outputs that you can get with the `ter
 
 ## Requirements
 ### Terraform plugins
-- [Terraform](https://www.terraform.io/downloads.html) 0.10.x
+- [Terraform](https://www.terraform.io/downloads.html) 0.12.x
```
Comment (Jberlinsky) [inline README.md]: We should add a note to the README along the lines of this one: https://github.com/terraform-google-modules/terraform-google-network/pull/55

Comment (ingwarr) [reply README.md]: Got it

**Comment by Jberlinsky** [comment]
Thanks @ingwarr! I'm going to work on https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/issues/230 so we can get CI working, then this can go for a final review pass.

**Comment by Jberlinsky** [comment]
@ingwarr Would you mind making an empty commit (`git commit --allow-empty -m 'CI rebuild'`) that triggers a CI rebuild? Something about the force-push is causing Concourse to error out:

```
get failed: failed to retrieve pull request: commit with ref '3278279985eba5ec286bbe98f86ab0f2ba4b01c5' does not exist
```

**Comment by ingwarr** [comment]
> Would you mind making an empty commit

No problem :)

**Comment by ingwarr** [comment]
@Jberlinsky @aaron-lane As I see it fails on integration test because needed variables are not present here https://github.com/ingwarr/cloud-foundation-toolkit/blob/master/infra/concourse/pipelines/terraform-google-vpn.yml
       `Error: Unassigned variable
       The input variable "mgt_project_id" has not been assigned a value. This is a
       bug in Terraform; please report it in a GitHub issue.
       Error: Unassigned variable
       The input variable "mgt_network" has not been assigned a value. This is a bug
       in Terraform; please report it in a GitHub issue.
       Error: Unassigned variable
       The input variable "prod_project_id" has not been assigned a value. This is a
       bug in Terraform; please report it in a GitHub issue.
       Error: Unassigned variable
       The input variable "prod_network" has not been assigned a value. This is a bug
       in Terraform; please report it in a GitHub issue.`
It works locally, it creates resources.
image
As I see `kitchen verify` option does not work because of empty https://github.com/ingwarr/terraform-google-vpn/blob/master/test/integration/single_tunnels/controls/gcp.rb

**Comment by morgante** [comment]
If new variables are needed, you should be capable of defining them in a pull request.

**Comment by ingwarr** [comment]
> If new variables are needed, you should be capable of defining them in a pull request.
@morgante I absolutely agree with you, I think that to test this module it is necessary to do dynamic creation of projects and networks in them. Otherwise, we will have to keep static objects for testing.

File: CHANGELOG.md [hunk 2]
```diff
@@ -18,3 +22,7 @@ project adheres to [Semantic Versioning](http://semver.org/).
 ## [0.1.0]
 ### ADDED
 - Initial module release.
+
+[0.1.0]: https://github.com/terraform-google-modules/terraform-google-vpn/releases/tag/v0.1.0
+[0.2.0]: https://github.com/terraform-google-modules/terraform-google-vpn/releases/tag/v0.2.0
+[#12]: https://github.com/terraform-google-modules/terraform-google-vpn/pull/12
```
Comment (bohdanyurov-gl) [inline CHANGELOG.md]: Wrong link to PR

File: main.tf [hunk 3]
```diff
@@ -15,56 +15,51 @@
...
+  count                     = var.cr_name != "" ? var.tunnel_count : 0
   name                      = "bgp-session-${count.index}"
-  router                    = "${var.cr_name}"
-  region                    = "${var.region}"
-  peer_ip_address           = "${element(var.bgp_remote_session_range, count.index)}"
-  peer_asn                  = "${element(var.peer_asn, count.index)}"
-  advertised_route_priority = "${var.advertised_route_priority}"
+  router                    = var.cr_name
+  region                    = var.region
+  peer_ip_address           = element(var.bgp_remote_session_range, count.index)
```
Comment (bohdanyurov-gl) [inline main.tf]: Please use "[count.index]"

File: main.tf [hunk 4]
```diff
@@ -15,56 +15,51 @@
...
+  peer_asn                  = element(var.peer_asn, count.index)
```
Comment (bohdanyurov-gl) [inline main.tf]: same

File: tunnel.tf [hunk 5]
```diff
@@ -20,42 +20,43 @@ resource "random_id" "ipsec_secret" {
...
-  name          = "${var.tunnel_count == 1 ? format("%s-%s",local.tunnel_name_prefix,"1") : format("%s-%d",local.tunnel_name_prefix,count.index+1)}"
-  region        = "${var.region}"
-  project       = "${var.project_id}"
-  peer_ip       = "${element(var.peer_ips, count.index)}"
-  shared_secret = "${local.default_shared_secret}"
+  count         = var.cr_name == "" ? var.tunnel_count : 0
+  name          = var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix, "1") : format("%s-%d", local.tunnel_name_prefix, count.index + 1)
+  region        = var.region
+  project       = var.project_id
+  peer_ip       = element(var.peer_ips, count.index)
```
Comment (bohdanyurov-gl) [inline tunnel.tf]: Please use "[count.index]"

File: tunnel.tf [hunk 6]
```diff
@@ -20,42 +20,43 @@ resource "random_id" "ipsec_secret" {
...
-  name          = "${var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix,"1") : format("%s-%d",local.tunnel_name_prefix,count.index+1)}"
-  region        = "${var.region}"
-  project       = "${var.project_id}"
-  peer_ip       = "${element(var.peer_ips, count.index)}"
-  shared_secret = "${local.default_shared_secret}"
+  count         = var.cr_name != "" ? var.tunnel_count : 0
+  name          = var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix, "1") : format("%s-%d", local.tunnel_name_prefix, count.index + 1)
+  region        = var.region
+  project       = var.project_id
+  peer_ip       = element(var.peer_ips, count.index)
```
Comment (bohdanyurov-gl) [inline tunnel.tf]: Please use "[count.index]"

File: variables.tf [hunk 7]
```diff
@@ -46,68 +46,74 @@ variable "local_traffic_selector" {
...
 variable "remote_subnet" {
-  description = "remote subnet ip range in CIDR format - x.x.x.x/x"
-  type        = "list"
-  default     = []
+description = "remote subnet ip range in CIDR format - x.x.x.x/x"
+type        = list(string)
+default     = []
 }
 
 variable "shared_secret" {
```
Comment (bohdanyurov-gl) [inline variables.tf]: missing type. Please also update other variables

File: variables.tf [hunk 8]
```diff
@@ -46,68 +46,74 @@ variable "local_traffic_selector" {
...
-  default     = ""
+description = "The name of cloud router for BGP routing"
+default     = ""
 }
 
 variable "peer_asn" {
-  type        = "list"
-  description = "Please enter the ASN of the BGP peer that cloud router will use"
-  default     = ["65101"]
+type        = list(string)
```
Comment (bohdanyurov-gl) [inline variables.tf]: Are you sure ASNs shouldn't be numbers?

**Comment by bohdanyurov-gl** [comment]
Please also squash all commits and rebase from latest master.

File: variables.tf (hunk 8 above)
Comment (morgante) [reply variables.tf]: We should treat them as strings.

File: CHANGELOG.md (hunk 2 above)
Comment (ingwarr) [reply CHANGELOG.md]: fixed

**Comment by ingwarr** [comment]
> Please also squash all commits and rebase from latest master.

@bohdanyurov-gl Done

File: Makefile [hunk 9]
```diff
@@ -21,7 +21,7 @@ SHELL := /usr/bin/env bash
 # Docker build config variables
 CREDENTIALS_PATH 			?= /cft/workdir/credentials.json
 DOCKER_ORG 				:= gcr.io/cloud-foundation-cicd
-DOCKER_TAG_BASE_KITCHEN_TERRAFORM 	?= 0.11.10_216.0.0_1.19.1_0.1.10
+DOCKER_TAG_BASE_KITCHEN_TERRAFORM 	?= 2.4.0
```
Comment (aaron-lane) [inline Makefile]: This should be 2.3.0 as 2.4.0 is in development.

File: test/integration/single_tunnels/inspec.yml [hunk 10]
```diff
@@ -2,4 +2,4 @@ name: single_tunnels
 depends:
   - name: inspec-gcp
     git: https://github.com/inspec/inspec-gcp.git
-    version: '~> 0.9.0'
+    version: '~> 0.11.0'
```
Comment (aaron-lane) [inline test/integration/single_tunnels/inspec.yml]: `version` is not actually supported by the git dependency.

```suggestion
    tag: v0.11.0
```

File: main.tf [hunk 11]
```diff
@@ -15,56 +15,51 @@
...
+  count                     = var.cr_name != "" ? var.tunnel_count : 0
   name                      = "bgp-session-${count.index}"
-  router                    = "${var.cr_name}"
-  region                    = "${var.region}"
-  peer_ip_address           = "${element(var.bgp_remote_session_range, count.index)}"
-  peer_asn                  = "${element(var.peer_asn, count.index)}"
-  advertised_route_priority = "${var.advertised_route_priority}"
+  router                    = var.cr_name
+  region                    = var.region
+  peer_ip_address           = element(var.bgp_remote_session_range, [count.index])
```
Comment (aaron-lane) [inline main.tf]: Why is `count.index` wrapped in a list? This is not valid syntax for `element`.

File: main.tf [hunk 12]
```diff
@@ -15,56 +15,51 @@
...
+  peer_asn                  = element(var.peer_asn, [count.index])
```
Comment (aaron-lane) [inline main.tf]: Same issue with `count.index`.

File: tunnel.tf [hunk 13]
```diff
@@ -20,42 +20,43 @@ resource "random_id" "ipsec_secret" {
...
-  name          = "${var.tunnel_count == 1 ? format("%s-%s",local.tunnel_name_prefix,"1") : format("%s-%d",local.tunnel_name_prefix,count.index+1)}"
-  region        = "${var.region}"
-  project       = "${var.project_id}"
-  peer_ip       = "${element(var.peer_ips, count.index)}"
-  shared_secret = "${local.default_shared_secret}"
+  count         = var.cr_name == "" ? var.tunnel_count : 0
+  name          = var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix, "1") : format("%s-%d", local.tunnel_name_prefix, count.index + 1)
+  region        = var.region
+  project       = var.project_id
+  peer_ip       = element(var.peer_ips, [count.index])
```
Comment (aaron-lane) [inline tunnel.tf]: Same issue with `count.index`.

File: tunnel.tf [hunk 14]
```diff
@@ -20,42 +20,43 @@ resource "random_id" "ipsec_secret" {
...
-  name          = "${var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix,"1") : format("%s-%d",local.tunnel_name_prefix,count.index+1)}"
-  region        = "${var.region}"
-  project       = "${var.project_id}"
-  peer_ip       = "${element(var.peer_ips, count.index)}"
-  shared_secret = "${local.default_shared_secret}"
+  count         = var.cr_name != "" ? var.tunnel_count : 0
+  name          = var.tunnel_count == 1 ? format("%s-%s", local.tunnel_name_prefix, "1") : format("%s-%d", local.tunnel_name_prefix, count.index + 1)
+  region        = var.region
+  project       = var.project_id
+  peer_ip       = element(var.peer_ips, [count.index])
```
Comment (aaron-lane) [inline tunnel.tf]: Same issue with `count.index`.

File: Makefile (hunk 9 above)
Comment (ingwarr) [reply Makefile]: Fixed

File: tunnel.tf (hunk 13 above)
Comment (ingwarr) [reply tunnel.tf]: Done

File: tunnel.tf (hunk 14 above)
Comment (ingwarr) [reply tunnel.tf]: Done

File: main.tf (hunk 11 above)
Comment (ingwarr) [reply main.tf]: Done

File: main.tf (hunk 12 above)
Comment (ingwarr) [reply main.tf]: Done

**Comment by ingwarr** [comment]
image

**Comment by ingwarr** [comment]
Tests failed d to:
       Error: Error creating instance: googleapi: Error 409: The resource 'projects/ci-vm-module/zones/us-central1-a/instances/umig-static-ips-001' already exists, alreadyExists
       
         on ../../../../modules/umig/main.tf line 43, in resource "google_compute_instance_from_template" "compute_instance":
         43: resource "google_compute_instance_from_template" "compute_instance" {
So we should delete it manually

**Pull request merged**